File: challenges/partitionLinkedList/Solutions/partitionLinkedList.js
function LinkedList() {
	this.head = null;
	this.tail = null;
}

function Node(data) {
	this.data = data;
	this.next = null;
}

LinkedList.prototype.addToHead = function(el) {
	if (el) {
	    var newNode = new Node(el);
		if(this.head === null){
			this.tail = this.head = newNode;
		}else{
			newNode.next = this.head;
			this.head = newNode;
		}
	} else {
		return undefined;
	}
}

function partition(node, x) {
    // create two linkedlists - small and large
    // create four pointers point to two linkedlists' heads and tails
    var smallHead = null;
    var smallTail = null;
    var largeHead = null;
    var largeTail = null;
    
    // traverse the linkedlist from the head node
    // add the node to small linkedlist if the node's data is less than x
    // add the node to large linkedlist if the node's data is larger than or equal to x
    while(node !== null) {
        var next = node.next;
        
        if (node.data < x) {
            if (smallHead === null) {
                smallHead = smallTail = node;
            } else {
                smallTail.next = node;
                smallTail = node;
                smallTail.next = null;
            }
        } else {
            if (largeHead === null) {
                largeHead = largeTail = node;
            } else {
                largeTail.next = node;
                largeTail = node;
                largeTail.next = null;
            }
        }
        node = next;
    }
    
    // merge two linkedlist
    if (smallHead) {
        smallTail.next = largeHead;
        return smallHead;   
    } else {
        return largeHead;
    }
}

// Test
var sll = new LinkedList();
sll.addToHead(3);
sll.addToHead(17);
sll.addToHead(2);
sll.addToHead(8);
sll.addToHead(35);
console.log('sll ', sll); // 35, 8, 2, 17, 3 
console.log('after partition: ', partition(sll.head, 9)); // 8, 2, 3, 35, 17
console.log('after partition: ', partition(sll.head, 1)); // 35, 8, 2, 17, 3
console.log('after partition: ', partition(sll.head, 36)); // 35, 8, 2, 17, 3

File: challenges/partitionLinkedList/Solutions/partitionLinkedList.py
# Submitted by bevdsouza


class Node(object):
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList(object):
    head = None

    def add_node(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
        else:
            curr = self.head
            while curr.next:
                curr = curr.next
            curr.next = new_node


def partition_list(linked_list, x):
    node = linked_list.head

    before_val = None
    after_val = None
    x_node = None

    while node:
        new_node = node.next
        if node.data < x:
            node.next = before_val
            before_val = node
        elif node.data > x:
            node.next = after_val
            after_val = node
        else:
            x_node = node
        node = new_node

    if x_node:
        x_node.next = after_val
        after_val = x_node

    if not before_val:
        return after_val

    head = before_val
    while before_val.next:
        before_val = before_val.next

    before_val.next = after_val
    return head


def print_list(li):
    while li:
        print li.data
        li = li.next

if __name__ == '__main__':
    l = LinkedList()
    l.add_node(4)
    l.add_node(5)
    l.add_node(12)
    l.add_node(42)
    l.add_node(6)
    l.add_node(23)
    l.add_node(11)
    l.add_node(24)
    new_head = partition_list(l, 12)
    print_list(new_head)
